src/os/bluestore/BlueStore.cc

```diff
@@ -3194,94 +3222,108 @@ int BlueStore::_do_read(
+    unsigned l_off = pos - lp->first;
+    unsigned b_off = l_off + lp->second.offset;
+    unsigned b_len = std::min(std::min(left,
+				       lp->second.length - l_off),
+			      bptr->blob.length - b_off);
+
+    ready_regions_t cache_res;
+    interval_set<uint64_t> cache_interval;
+    bptr->bc.read(b_off, b_len, cache_res, cache_interval);
+    dout(20) << __func__ << "  blob " << *bptr << std::hex
+	     << " need 0x" << b_off << "~" << b_len
+	     << " cache has 0x" << cache_interval
+	     << std::dec << dendl;
+
+    auto pc = cache_res.begin();
+    while (b_len > 0) {
+      unsigned l;
+      if (pc != cache_res.end() &&
+	  pc->first == b_off) {
+	l = pc->second.length();
+	ready_regions[pos].claim(pc->second);
+	dout(30) << __func__ << "    use cache 0x" << std::hex << pos << ": 0x"
+		 << b_off << "~" << l << std::dec << dendl;
+	++pc;
+      } else {
+	l = b_len;
+	if (pc != cache_res.end()) {
+	  assert(pc->first > b_off);
+	  l = pc->first - b_off;
+	}
+	dout(30) << __func__ << "    will read 0x" << std::hex << pos << ": 0x"
+		 << b_off << "~" << l << std::dec << dendl;
+	blobs2read[bptr].emplace_back(region_t(pos, b_off, l));
       }
-      ++lext;
-    } else if (off >= lext->first + lext->second.length) {
-      //handling the case when the first lookup get into the previous block due to the hole
-      l2read = 0;
-      ++lext;
-    } else {
-      //hole found
-      l2read = MIN(l, lext->first - off);
+      pos += l;
+      b_off += l;
+      left -= l;
+      b_len -= l;
     }
-    off += l2read;
-    l -= l2read;
+    ++lp;
   }
 
   //enumerate and read/decompress desired blobs
   blobs2read_t::iterator b2r_it = blobs2read.begin();
   while (b2r_it != blobs2read.end()) {
-    const bluestore_blob_t* bptr = b2r_it->first;
-    regions2read_t r2r = b2r_it->second;
-    regions2read_t::const_iterator r2r_it = r2r.cbegin();
-    if (bptr->has_flag(bluestore_blob_t::FLAG_COMPRESSED)) {
+    Blob* bptr = b2r_it->first;
+    dout(20) << __func__ << "  blob " << *bptr << std::hex
+	     << " need 0x" << b2r_it->second << std::dec << dendl;
+    if (bptr->blob.has_flag(bluestore_blob_t::FLAG_COMPRESSED)) {
       bufferlist compressed_bl, raw_bl;
-
-      int r = _read_whole_blob(bptr, o, &compressed_bl);
-      if (r < 0)
-	return r;
-      if (bptr->csum_type != bluestore_blob_t::CSUM_NONE) {
-	r = _verify_csum(bptr, 0, compressed_bl);
-	if (r < 0) {
-	  dout(20) << __func__ << "  blob reading 0x" << std::hex
-		   << r2r_it->logical_offset
-		   << " 0~" << compressed_bl.length()
-		   << std::dec << " csum verification failed." << dendl;
-	  return -EIO;
-	}
+      IOContext ioc(NULL);   // FIXME?
+      bptr->blob.map(
+	0, bptr->blob.get_ondisk_length(),
+	[&](uint64_t offset, uint64_t length) {
+	  bufferlist t;
+	  int r = bdev->read(offset, length, &t, &ioc, false);
+	  assert(r == 0);
+	  compressed_bl.claim_append(t);
+	});
+      if (bptr->blob.csum_type != bluestore_blob_t::CSUM_NONE &&
```

Comment:
we can get rid off the check for CSUM_NONE as blob_t::verify_csum does that by itself